### DISCUSSION TITLE
Dynamic Current

### DISCUSSION THREAD

**Discussion by TazerReloaded** [CAP1Sup/Intellistep#13] [Ideas]
Hello,

last night I had an idea for automatic current adjustment to prevent overheating motor and drivers even with high current settings.
Some firmwares already implement an extra setting for holding current, because when the motor stands still, the current can be drastically reduced while still keeping position. I think because the S42B is highly configurable and knows motor position and speed at all times, we can expand on this concept.

My idea is a dynamic current control, according to a linear or maybe quadratic equation, based on two input values and two configuration factors (0 to 1) as well as two limits:
- Motor speed, can be derived from the incoming step pulses. More speed, more current.
- Motor speed factor, a setting which controls how dependent on the speed the current is
- Motor acceleration, can be derived from the change in speed over time. More acceleration, more current.
- Motor acceleration factor, a setting which controls how dependent on the acceleration the current is
- Configuration value for minimum current limit (holding torque)
- Configuration value for maximum current limit (at max speed or max acceleration)

I considered the following practical applications:
- X/Y-Axis motor (cartesian):
  - High torque needed for acceleration (getting the gantry up to speed), low torque for keeping speed or standing still (inertia)
  - Factors could be e.g. 1 for acceleration, 0.5 for speed
- Extruder motor:
  - High torque needed for speed (pushing filament), low torque for acceleration (very low mass of the gears)
  - Factors could be e.g. 0.5 for acceleration, 1 for speed
- Z-Axis motor (cartesian, threaded rod):
  - Neither accelerating nor running the motor need any huge amounts of torque, but position should be kept against the weight of the gantry or bed
  - Factors could be e.g. 0.2 for acceleration and 0.2 for speed

The controller would then automatically adjust motor current based on demand, and I think we could even write an auto-tuning algorithm for finding the factors automatically based on the hall sensor readings in a calibration sequence. I don't know how fast the current settings of the drivers can be changed, that could be a limiting factor.

Let me know what you think about this concept!

**Comment by CAP1Sup** [comment]
This sounds like an excellent concept! If we have the ability to do it, why not? For now, I'd like to focus on just getting the encoder speed working and the coils driven correctly. However, after that is done, this will be the first thing on my list to implement. Also, now that we have the temperature of the encoder, we can reduce the current of the motor if we have temperatures over a specific value, say 80C. The potential of this board is really great, it just needs the firmware to do it. As for the current setting, that is done through a PWM pin. So really, we can change it as fast as we want. I'll add this idea to the ToDo list. Definitely a good idea though!

**Reply by CAP1Sup** [reply]
@TazerReloaded
I began the implementation of your idea. Currently I only have acceleration and idle current factors, but I'll add the speed current factor tonight. I still have a ton of work to do so that the display and parser are able to understand the new feature. I also added an estimation of acceleration, so hopefully that works. I haven't been able to test it yet. If you wanted to check out the progress, I'm going to develop it on the betterReverse branch, then merge it once I finish it. Let me know what you think, and if it matches with your original vision. This is the type of feature that can set Intellistep apart from traditional stepper drivers.
